### PR TITLE
gh-110805: Allow the repl to show source code and complete tracebacks

### PR DESCRIPTION
**Pull request by pablogsal**
* Issue: gh-110805

### CODE REVIEW

File: Python/pythonrun.c [hunk 1]
```diff
@@ -1262,13 +1264,46 @@ run_eval_code_obj(PyThreadState *tstate, PyCodeObject *co, PyObject *globals, Py
...
-            PyCompilerFlags *flags, PyArena *arena)
+            PyCompilerFlags *flags, PyArena *arena, PyObject* interactive_src)
 {
     PyThreadState *tstate = _PyThreadState_GET();
+    PyObject* interactive_filename = NULL;
+    if (interactive_src) {
+        PyInterpreterState *interp = tstate->interp;
+        interactive_filename = PyUnicode_FromFormat("<python-input-%d>", interp->_interactive_src_count++);
+        // TODO: Maybe we are leaking
+        filename = interactive_filename;
```
Comment (lysnikolaou) [inline Python/pythonrun.c]: We are probably leaking here.

File: Lib/traceback.py [hunk 2]
```diff
@@ -331,7 +331,8 @@ def line(self):
         if self._line is None:
             if self.lineno is None:
                 return None
-            self._line = linecache.getline(self.filename, self.lineno)
+            if self._line is None:
+                self._line = linecache.getline(self.filename, self.lineno)
```
Comment (lysnikolaou) [inline Lib/traceback.py]: Is this needed?

Comment (pablogsal) [reply Lib/traceback.py]: 👍

**Review by lysnikolaou — APPROVED** [review]
Looks good! Great one! 🚀

**Comment by lysnikolaou** [comment]
Some lint errors still there.

**Comment by merwok** [comment]
Which version is this for? (you can answer by adding label)

**Comment by pablogsal** [comment]
3.13

**Pull request merged**